canonical/hydra-operator
# Initial deployment

This PR introduces initial setup of the hydra charm.

# Added

- pebble container for hydra
- relation with postgresql-k8s
- upgraded the initial charm with hydra v2.0.2

Note that some configs like `--dev` will be removed at later phase of development.

# Testing
1. Deploy postgresql-k8s charm:
`juju deploy postgresql-k8s --channel edge --trust`
3. Pack and deploy hydra:
`juju deploy ./hydra*.charm --resource oci-image=$(yq eval '.resources.oci-image.upstream-source' metadata.yaml)`
`juju relate postgresql-k8s hydra`
4. Confirm that the charm gets active and the containers are ready.
5. To test hydra functionalities, execute the container:
`kubectl exec -it hydra-0 -c hydra -n <model> -- sh`
Create an exemplary client (attention: the CLI and API have changed in v2):
```
# hydra create client --endpoint http://127.0.0.1:4445/ --name example-client
CLIENT ID	b55b6857-968e-4fb7-be77-f701ec751405	
CLIENT SECRET	b3wFYH2N_epJY6C8jCuinBRP60		
GRANT TYPES	authorization_code			
RESPONSE TYPES	code					
SCOPE		offline_access offline openid		
AUDIENCE						
REDIRECT URIS	
```
Confirm creation by listing the clients:
```
# hydra list clients --endpoint http://127.0.0.1:4445/
CLIENT ID				CLIENT SECRET	GRANT TYPES		RESPONSE TYPES	SCOPE				AUDIENCE	REDIRECT URIS	
b55b6857-968e-4fb7-be77-f701ec751405			authorization_code	code		offline_access offline openid					
					
NEXT PAGE TOKEN					
IS LAST PAGE				true
```

# Next steps

- [x] Add hydra-maester container to manage oauth2clients resources [will be managed through juju/UI]
- [x] Relate hydra to postgresql
- [x] Add sql migration job
